server.ts

```diff
@@ -201,62 +212,191 @@ export function app() {
   }
 }
 
+/**
+ * Render page content on server side using Angular SSR. By default this page content is
+ * returned to the user.
+ * @param req current request
+ * @param res current response
+ * @param sendToUser if true (default), send the rendered content to the user.
+ * If false, then only save this rendered content to the in-memory cache (to refresh cache).
+ */
+function serverSideRender(req, res, sendToUser: boolean = true) {
+  // Render the page via SSR (server side rendering)
+  res.render(indexHtml, {
+    req,
+    res,
+    preboot: environment.universal.preboot,
+    async: environment.universal.async,
+    time: environment.universal.time,
+    baseUrl: environment.ui.nameSpace,
+    originUrl: environment.ui.baseUrl,
+    requestUrl: req.originalUrl,
+    providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }]
+  }, (err, data) => {
+    if (hasNoValue(err) && hasValue(data)) {
+      res.locals.ssr = true;  // mark response as SSR (enables text compression)
+      // save server side rendered data to cache
+      saveToCache(req, data);
+      if (sendToUser) {
+        // send rendered page to user
+        res.send(data);
+      }
+    } else if (hasValue(err) && err.code === 'ERR_HTTP_HEADERS_SENT') {
+      // When this error occurs we can't fall back to CSR because the response has already been
+      // sent. These errors occur for various reasons in universal, not all of which are in our
+      // control to solve.
+      console.warn('Warning [ERR_HTTP_HEADERS_SENT]: Tried to set headers after they were sent to the client');
+    } else {
+      console.warn('Error in server-side rendering (SSR)');
+      if (hasValue(err)) {
+        console.warn('Error details : ', err);
+      }
+      if (sendToUser) {
+        console.warn('Falling back to serving direct client-side rendering (CSR).');
+        clientSideRender(req, res);
+      }
+    }
+  });
+}
+
+/**
+ * Send back response to user to trigger direct client-side rendering (CSR)
+ * @param req current request
+ * @param res current response
+ */
+function clientSideRender(req, res) {
+  res.render(indexHtml, {
+    req,
+    providers: [{
+      provide: APP_BASE_HREF,
+      useValue: req.baseUrl
+    }]
+  });
+}
+
+
 /*
- * Adds a cache control header to the response
- * The cache control value can be configured in the environments file and defaults to max-age=60
+ * Adds a Cache-Control HTTP header to the response.
+ * The cache control value can be configured in the config.*.yml file
+ * Defaults to max-age=604,800 seconds (1 week)
  */
-function cacheControl(req, res, next) {
+function addCacheControl(req, res, next) {
   // instruct browser to revalidate
-  res.header('Cache-Control', environment.cache.control || 'max-age=60');
+  res.header('Cache-Control', environment.cache.control || 'max-age=604800');
   next();
 }
 
+/*
+ * Enable server-side caching of pages rendered via SSR.
+ */
+function enableCache() {
+  if (cacheEnabled()) {
```

Comment:
This looks a bit confusing, would be more clear if the method was called something like `initCache`

Reply:
Good point.  Will rename to `initCache()`